ipa-core/src/protocol/context/mod.rs

```diff
@@ -312,7 +316,35 @@ impl<'a, B: ShardBinding> Context for Base<'a, B> {
 
 /// Context for MPC circuits that can operate on multiple shards. Provides access to shard information
 /// via [`ShardConfiguration`] trait.
-pub trait ShardedContext: Context + ShardConfiguration {}
+pub trait ShardedContext: Context + ShardConfiguration {
+    /// Requests data to be received from all shards that are registered in the system.
+    /// Shards that don't have any data to send, must explicitly open and close the send channel.
+    fn recv_from_shards<M: Message>(
+        &self,
+    ) -> impl Stream<Item = (ShardIndex, Result<M, crate::error::Error>)> + Send {
+        stream::select_all(
+            self.peer_shards()
+                .map(|origin| self.shard_recv_channel(origin).map(move |v| (origin, v))),
+        )
+    }
+
+    /// Picks a shard according to the value obtained from sampling PRSS shared with the given helper.
+    fn pick_shard(&self, record_id: RecordId, direction: Direction) -> ShardIndex {
+        // FIXME: update PRSS trait to compute only left or right part
```

Comment:
My initial reaction was that this isn't that hard (maybe something like `impl FromPrss for Left<T>`), but upon going to try and mock it up, I realized it's a bit harder than that because it needs to propagate through a few layers of functions in the PRSS implementation.

Since the PRSS stuff needs to be monomorphized anyways, maybe the compiler will figure out that it can skip half of the PRSS generation?

Reply:
my worry is that as you said there are quite a few layers before we get to AES, so compiler may not have enough memory to keep track of things and/or time to do that. Cutting of 50% of CPU time to generate these masks seems to be important enough to have some guarantee of it happening